kotlin/reflect/src/main/java/com/squareup/moshi/kotlin/reflect/KotlinJsonAdapter.kt

```diff
@@ -79,25 +82,38 @@ internal class KotlinJsonAdapter<T>(
             "Multiple values for '${binding.property.name}' at ${reader.path}")
       }
 
-      values[index] = binding.adapter.fromJson(reader)
+      values[index] = when (failureStrategy) {
+        FailureStrategy.FAIL_FAST -> binding.adapter.fromJson(reader)
+        FailureStrategy.ACCUMULATE_ERRORS -> binding.adapter.nullSafe().fromJson(reader)
+      }
 
       if (values[index] == null && !binding.property.returnType.isMarkedNullable) {
-        throw JsonDataException(
-            "Non-null value '${binding.property.name}' was null at ${reader.path}")
+        when(failureStrategy) {
+            FailureStrategy.FAIL_FAST -> throw JsonDataException(
+                    "Non-null value '${binding.property.name}' was null at ${reader.path}")
+            FailureStrategy.ACCUMULATE_ERRORS -> failures!!.add("'${binding.property.name}'")
+        }
       }
     }
     reader.endObject()
 
+
     // Confirm all parameters are present, optional, or nullable.
     for (i in 0 until constructorSize) {
       if (values[i] === ABSENT_VALUE && !constructor.parameters[i].isOptional) {
         if (!constructor.parameters[i].type.isMarkedNullable) {
-          throw JsonDataException(
-              "Required value '${constructor.parameters[i].name}' missing at ${reader.path}")
+          when(failureStrategy){
+            FailureStrategy.FAIL_FAST -> throw JsonDataException(
+                    "Required value '${constructor.parameters[i].name}' missing at ${reader.path}")
+            FailureStrategy.ACCUMULATE_ERRORS -> failures!!.add("'${constructor.parameters[i].name}'")
+          }
         }
         values[i] = null // Replace absent with null.
       }
     }
+    if (failureStrategy == FailureStrategy.ACCUMULATE_ERRORS && failures!!.isNotEmpty()) {
+      throw JsonDataException("Required value(s) ${failures.joinToString(separator = " ")} missing or null at ${reader.path}")
```

Comment:
Non-nullable null values and missing fields in the same exception. Not sure if it's worth it to separate them?

Reply:
Put the cause of the failure into the message list and remove "missing or null" from here